CIP-0022/CIP-0022.md

```diff
@@ -0,0 +1,139 @@
+println("domain: $domain, nonce: ${nonce.toHexString()}")
+```
+
+## Code Example (Pool Operator side):
+
+```kotlin
+// Node operational VRF-Verification-Key: pool.vrf.vkey
+//{
+//    "type": "VrfVerificationKey_PraosVRF",
+//    "description": "VRF Verification Key",
+//    "cborHex": "5820e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381"
+//}
+//
+// Node operational VRF-Signing-Key: pool.vrf.skey
+//{
+//    "type": "VrfSigningKey_PraosVRF",
+//    "description": "VRF Signing Key",
+//    "cborHex": "5840adb9c97bec60189aa90d01d113e3ef405f03477d82a94f81da926c90cd46a374e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381"
+//}
+
+// We assume the pool operator has access to the pool's vrf secret key
+val skeyCbor = "5840adb9c97bec60189aa90d01d113e3ef405f03477d82a94f81da926c90cd46a374e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381".hexToByteArray()
+val vrfSkey = (CborReader.createFromByteArray(skeyCbor).readDataItem() as CborByteString).byteArrayValue()
+val vkeyCbor = "5820e0ff2371508ac339431b50af7d69cde0f120d952bb876806d3136f9a7fda4381".hexToByteArray()
+val vrfVkey = (CborReader.createFromByteArray(vkeyCbor).readDataItem() as CborByteString).byteArrayValue()
+
+// Client side, construct and sign the challenge
+val challengeSeed = "cip-0022${domain}".toByteArray() + nonce
+val challenge = SodiumLibrary.cryptoBlake2bHash(challengeSeed, null)
+println("challenge: ${challenge.toHexString()}")
+
+val signature = SodiumLibrary.cryptoVrfProve(vrfSkey, challenge)
+println("signature: ${signature.toHexString()}")
+```
+
+## Code Example (Validating server):
+
+```kotlin
+// Server side, verify the message based on only knowing the pool_id, public vkey, signature, and constructing
+// the challenge ourselves the same way the client should have.
+val challengeSeed = "cip-0022${domain}".toByteArray() + nonce
+val challenge = SodiumLibrary.cryptoBlake2bHash(challengeSeed, null)
+
+// Get the vkeyHash for a pool from the "query pool-params" cardano-cli command
+// This comes from the pool's registration certificate on the chain.
+val vkeyHash = "f58bf0111f8e9b233c2dcbb72b5ad400330cf260c6fb556eb30cefd387e5364c".hexToByteArray()
+
+// Verify that the vkey from the latest minted block on the blockchain (or the client supplied if they
+// haven't yet minted a block) is the same as the one on-chain in the pool's registration certificate
+val vkeyHashVerify = SodiumLibrary.cryptoBlake2bHash(vrfVkey, null)
+assertThat(vkeyHash).isEqualTo(vkeyHashVerify)
+
+// Verify that the signature is a valid format. This will fail if the signature is mal-formed
+val signatureHash = SodiumLibrary.cryptoVrfProofToHash(signature)
+println("signatureHash: ${signatureHash.toHexString()}")
+
+// Verify that the signature matches
+val verification = SodiumLibrary.cryptoVrfVerify(vrfVkey, signature, challenge)
+println("verification: ${verification.toHexString()}")
+assertThat(signatureHash).isEqualTo(verification)
```

Comment:
Why are we performing this equality check? In libsodium, the output of `cryptoVrfVerify(vrfVkey, signature, challenge)` is `cryptoVrfProofToHash(signature)` (see [here](https://github.com/input-output-hk/libsodium/blob/draft-irtf-cfrg-vrf-03/src/libsodium/crypto_vrf/ietfdraft03/verify.c#L180)). This is, assuming that the signature is valid, so this check seems redundant. It is sufficient to ensure that `cryptoVrfVerify` does not output an error. There is no need to compute `signatureHash` above.

Reply:
I've updated to remove the `cryptoVrfProofToHash(signature)` block of code and remove the final assertion that they match. You're correct that we can consider the signature verified as long as `cryptoVrfVerify` does not return an error.